src/main/scala/com/microsoft/azure/iot/iothubreact/sinks/OffsetCommitSink.scala

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+package com.microsoft.azure.iot.iothubreact.sinks
+
+import java.util.concurrent.CompletionStage
+import java.util.concurrent.atomic.AtomicLong
+
+import akka.Done
+import akka.japi.function.Procedure
+import akka.stream.javadsl.{Sink ⇒ JavaSink}
+import akka.stream.scaladsl.{Sink ⇒ ScalaSink}
+import com.microsoft.azure.iot.iothubreact.checkpointing.OffsetLoader
+import com.microsoft.azure.iot.iothubreact.checkpointing.backends.CheckpointBackend
+import com.microsoft.azure.iot.iothubreact.config.{Configuration, IConfiguration}
+import com.microsoft.azure.iot.iothubreact.{Logger, MessageFromDevice}
+
+import scala.collection.concurrent.TrieMap
+import scala.concurrent.Future
+
+final case class OffsetCommitSink(parallelism: Int, backend: CheckpointBackend, config: IConfiguration) extends ISink[MessageFromDevice] with Logger {
+
+  val current: TrieMap[Int, Long] = TrieMap()
+  OffsetLoader.GetSavedOffsets(config).foreach{ case (a, c) ⇒
+    current += a → c.toLong
+  }
+
+  private[this] object JavaSinkProcedure extends Procedure[MessageFromDevice] {
+    @scala.throws[Exception](classOf[Exception])
+    override def apply(m: MessageFromDevice): Unit = {
+      doWrite(m)
+    }
+  }
+
+  def scalaSink(): ScalaSink[MessageFromDevice, scala.concurrent.Future[Done]] = {
+    import scala.concurrent.ExecutionContext.Implicits.global
+    ScalaSink.foreachParallel[MessageFromDevice](parallelism) { doWrite }
+  }
+
+  private def doWrite(m: MessageFromDevice) = {
```

Comment:
Looks like this sink is updating the offset for every message in the stream. That's quite expensive and could impact the overall performance. 
This behavior should be similar to the out-of-process cp, i.e. reusing `checkpointing.frequency` and `checkpointing.countThreshold` to decide when it's time to write. A user can opt for the most expensive behavior with countThreshold set to 0